### ISSUE TITLE
Cross-Build For Android

### ISSUE THREAD

**Issue by cydhaselton** [cydhaselton/coreclr#2]
Referencing [this thread](https://github.com/cydhaselton/coreclr/issues/1#issuecomment-272668911) and tagging @janvorli, @jkotas and @qmfrederik as they participated in original.

I've set up an Ubuntu 14.04 server, installed coreclr prerequisites and the Android NDK toolchain. Before I start, are there any coreclr-specific instructions for cross-building?

**Comment by cydhaselton** [comment]
Currently cross-compiling lldb dependencies for build.

**Comment by janvorli** [comment]
@cydhaselton the way that Android NDK handles cross building seems different from how we do it for Linux. As for Linux, we build a crossfs, which is a folder containing headers, libraries and other files that are necessary for the build and are the same as if the build was running on the target machine. The root folder of the rootfs corresponds to the root folder of the target. For tools, we require installation of the binutils-aarch64-linux-gnu package.

Android NDK has a different structure. It has toolchains folder for the build tools and platforms folder for the libs / headers.  

I think that we cab try to use the binutils-aarch64-linux-gnu package can be used for building the stuff for now to minimize the changes to the cross build scripts.

The platforms should correspond to the ROOTFS_DIR, so we can use it for building:
`ROOTFS_DIR=/home/janvorli/ndk/android-ndk-r14-beta1/platforms/android-24/arch-arm64/ ./build.sh arm64 cross
`

When I try it, it fails because it cannot find some headers for packages that are obviously not part of the NDK. The one I hit was libicu-dev. 
For now, it seems that the way to go would be to gather / self-build all the missing dependencies for Android and then copy them to the proper subfolders of the ndk/android-ndk-r14-beta1/platforms/android-24/arch-arm64.

**Comment by cydhaselton** [comment]
@janvorli The NDK has a utility that builds a toolchain with the appropriate binaries and a sysroot with the Android-specific necessary headers and libs. The tool is called `make-standalone-toolchain.sh` and it's located in $NDK/build/tools.  You can run `make-standalone-toolchain.sh --help` to see the various options.  I suggest scrapping the beta and downloading the latest stable release (13B). 

Regarding dependencies; I copied over the libicu lib and headers to the appropriate locations in the toolchain sysroot. The next error I run into involves missing lldb; for some reason this wasn't an issue when building on device but it is now.  Will post the error after I finish re-building dev server.

**Comment by janvorli** [comment]
You can disable building the lldb plugin using the ENABLE_LLDBPLUGIN cmake variable. You can pass it to build.sh like this:
`cmakeargs -DENABLE_LLDBPLUGIN=0`

**Comment by janvorli** [comment]
You'll likely need to delete the bin folder first, otherwise cmake would not pick the variable setting and use the cached value instead.

**Comment by cydhaselton** [comment]
@janvorli How did you specify the target triplet in `cmakeargs`?  I've tried `-DTARGET` and `-DCMAKE_TARGET` neither of which the build picks up.

**Comment by janvorli** [comment]
I did not specify anything using that way. I have just asked a cross build targetting arm64:
`./build.sh arm64 cross` with the ROOTFS_DIR set to the root fs. That should be enough, I believe.

**Comment by cydhaselton** [comment]
@janvorli Odd. Running a modified version of the command you posted results in this: http://pastebin.com/svwg1Mu7

**Comment by janvorli** [comment]
@cydhaselton Have you installed the binutils-aarch64-linux-gnu package? And have you set the ROOTFS_DIR to something like what I had in my example above?

**Comment by cydhaselton** [comment]
@janvorli I didn't…did you see my reply to your post about the NDK toolchain? It's what I've installed.

I'll try setting ROOTFS_DIR before running ./buildsh

**Comment by janvorli** [comment]
Our cross build is not configured to work with the NDK toolchain, I think it would be easier to start with the toolchain from the package instead of starting modifying the cross build stuff to be able to use the NDK toolchain.

**Comment by janvorli** [comment]
But if you haven't set the ROOTFS_DIR before, maybe the toolchain from the NDK would actually work.

**Comment by cydhaselton** [comment]
I think I'm missing something. How would the `binutils-aarch64-linux-gnu ` package be able to generate a coreclr that works on Android?

The NDK script builds a toolchain and rootfs installation; wouldn't that be the same as installing the `binutils-aarch64-linux-gnu`?

**Comment by janvorli** [comment]
The binary tools for Android and Linux aarch64 should produce equivalent files, I believe. For Linux, we both install the binutils-aarch64-linux-gnu package and create ROOTFS. These two are orthogonal things. The binutils-aarch64-linux-gnu installs the tools to /usr and /usr/lib.

**Comment by cydhaselton** [comment]
The NDK installs the tools to $DIR/bin and $DIR/lib, and creates the rootfs in $DIR/sysroot.  Couldn't this setup be used for the build?

Also, to clarify, are you talking about using the `binutils-aarch64-linux-gnu` package with the NDK script generated sysroot?

**Comment by janvorli** [comment]
I believe the NDK tools can be used, but I was thinking that you'd need to modify the cross build scripts in coreclr and corefx. While with using the binutils-aarch64-linux-gnu package with the NDK script generated sysroot I believe should work fine without the need to modify the cross build scripts. But I may be wrong.

**Comment by cydhaselton** [comment]
@janvorli Similar errors with the `binutils-aarch64-linux-gnu` package: http://pastebin.com/0FYrBgeJ

**Comment by janvorli** [comment]
It seems there is something wrong with your ROOTFS_DIR. What do you see if you `ls` the path that you set the ROOTFS_DIR to? 
The one I used looks like this:
```
ls /home/janvorli/ndk/android-ndk-r14-beta1/platforms/android-24/arch-arm64/
usr

ls /home/janvorli/ndk/android-ndk-r14-beta1/platforms/android-24/arch-arm64/usr/
include  lib

ls /home/janvorli/ndk/android-ndk-r14-beta1/platforms/android-24/arch-arm64/usr/lib/
crtbegin_dynamic.o  crtend_android.o  libc.a            libdl.so         libGLESv2.so       liblog.so       libm.so          libstdc++.a   libz.a
crtbegin_so.o       crtend_so.o       libcamera2ndk.so  libEGL.so        libGLESv3.so       libm.a          libOpenMAXAL.so  libstdc++.so  libz.so
crtbegin_static.o   libandroid.so     libc.so           libGLESv1_CM.so  libjnigraphics.so  libmediandk.so  libOpenSLES.so   libvulkan.so

 ls /home/janvorli/ndk/android-ndk-r14-beta1/platforms/android-24/arch-arm64/usr/include/
alloca.h     byteswap.h  elf.h       fnmatch.h  grp.h       link.h    memory.h    OMXAL       resolv.h     SLES         sys          uchar.h     wchar.h
android      camera      endian.h    fts.h      inttypes.h  linux     mntent.h    pathconf.h  sched.h      stdatomic.h  syslog.h     ucontext.h  wctype.h
ar.h         complex.h   err.h       ftw.h      jni.h       locale.h  net         paths.h     search.h     stdint.h     termio.h     unistd.h    xlocale.h
arpa         ctype.h     errno.h     getopt.h   KHR         machine   netdb.h     poll.h      semaphore.h  stdio.h      termios.h    util.h      zconf.h
asm          dirent.h    fcntl.h     GLES       lastlog.h   malloc.h  netinet     pthread.h   setjmp.h     stdlib.h     thread_db.h  utime.h     zlib.h
asm-generic  dlfcn.h     features.h  GLES2      libgen.h    math.h    netpacket   pwd.h       sgtty.h      string.h     time64.h     utmp.h
assert.h     EGL         fenv.h      GLES3      limits.h    media     nsswitch.h  regex.h     signal.h     strings.h    time.h       vulkan
```

**Comment by janvorli** [comment]
I have forgotten to add that I have set the ROOTFS_DIR to /home/janvorli/ndk/android-ndk-r14-beta1/platforms/android-24/arch-arm64/

**Comment by cydhaselton** [comment]
I see the same thing, with the exception of having used the `make-standalone-toolchain,sh` to generate the headers and libs in the sysroot instead of directly using the platform dirs.  Generating the toolchain package is the recommended way to build the sysroot…see this: https://developer.android.com/ndk/guides/standalone_toolchain.html

The sysroot I generated targets android-22, which matches my device android version. I'm using r13b of the NDK, which is the latest stable version.

> I have forgotten to add that I have set the ROOTFS_DIR to /home/janvorli/ndk/android-ndk-r14-beta1/platforms/android-24/arch-arm64/

See previous posts about Android-recommended use of NDK toolchains

**Comment by janvorli** [comment]
Let me try to use the script too and see what happens on my machine.

**Comment by janvorli** [comment]
Hmm, using the generated rootfs doesn't work for me either. Guessing from the log, we need to modify the  cross/arm64/toolchain.cmake so that paths in it correspond to the ones in the  rootfs.

**Comment by cydhaselton** [comment]
What options did you use to generate the toolchain?
Additionally, shouldn't the cross toolchain cmake behave in the same way as the [CMAKE_FIND_ROOT_PATH](http://www.vtk.org/Wiki/CMake_Cross_Compiling) cmake variable, and append paths accordingly?

**Comment by janvorli** [comment]
Ok, if you set the ROOTFS_DIR to the sysroot subfolder of the generated toolchain dir, then you can get as far as I did before with the other root fs, building using the installed binutils-aarch64-linux-gnu.
If we want to use the clang and other tools provided in the NDK, then the toolchain.cmake and possibly something else will need to be modified to reflect the paths differences and use the compiler etc.

**Comment by cydhaselton** [comment]
@janvorli What env variables are you using to specify the binutils package?  I've installed it and the build still fails. In particular, the build is trying to link with the following:
`-target aarch64-linux-gnu -B /ndk/sysroot/usr/lib/gcc/aarch64-linux-gnu -L/ndk/sysroot/lib/aarch64-linux-gnu --sysroot=/ndk/sysroot`

**Comment by janvorli** [comment]
THere's no setting needed. using the "cross" option of the build.sh does all the settings. Look at the cross/arm64/toolchain.cmake, it all happens there.
The
```
set(TOOLCHAIN "aarch64-linux-gnu")
set(TOOLCHAIN_PREFIX ${TOOLCHAIN}-)
```
result in using aarch64-linux-gnu-ld and other tools instead of just ld,ar, etc.

**Comment by janvorli** [comment]
Btw here is what I did to use the installed binutils package:
```bash
mkdir ~/ndk
cd ~/ndk
wget https://dl.google.com/android/repository/android-ndk-r13b-linux-x86_64.zip
unzip android-ndk-r13b-linux-x86_64.zip
./android-ndk-r13b/build/tools/make_standalone_toolchain.py --arch arm64 --api 23 --install-dir ./toolchain
cd ~/git/coreclr
ROOTFS_DIR=/home/janvorli/ndk/toolchain/sysroot ./build.sh arm64 cross
```

**Comment by cydhaselton** [comment]
@janvorli I'm doing the exact same thing, and the build is using the correct binutils, but for some reason the build is appending the `TOOLCHAIN_PREFIX` to the sysroot path, which will fail.

**Comment by cydhaselton** [comment]
Here's the command I'm using: `ROOTFS_DIR=/ndk/sysroot ./build.sh arm64 cross`

**Comment by janvorli** [comment]
Can you please get me the build log from bin/obj/Linux.arm64.Debug/CMakeFiles/CMakeOutput.log?

**Comment by cydhaselton** [comment]
@janvorli Will do, but I think this is the problem:
```
set(CROSS_LINK_FLAGS "${CROSS_LINK_FLAGS} -target ${TOOLCHAIN}")                            set(CROSS_LINK_FLAGS "${CROSS_LINK_FLAGS} -B ${CROSS_ROOTFS}/usr/lib/gcc/${TOOLCHAIN}")     set(CROSS_LINK_FLAGS "${CROSS_LINK_FLAGS} -L${CROSS_ROOTFS}/lib/${TOOLCHAIN}")              set(CROSS_LINK_FLAGS "${CROSS_LINK_FLAGS} --sysroot=${CROSS_ROOTFS}")
```
How is your build finding the crti.o files if the link flag adds `-L/path/to/sysroot/usr/lib/gcc/aarch64-linux-gnu`?

**Comment by cydhaselton** [comment]
@janvorli CMakeOutput.log: http://pastebin.com/YwSwdC01

**Comment by janvorli** [comment]
I guess it is using the ones from `/usr/aarch64-linux-gnu/lib/crti.o` instead. The -L options is just additive and if it contains non-existent path, it doesn't care if it can find the files elsewhere. It works even if I comment out the `set(CROSS_LINK_FLAGS "${CROSS_LINK_FLAGS} -L${CROSS_ROOTFS}/lib/${TOOLCHAIN}")`

**Comment by janvorli** [comment]
@cydhaselton I don't see any error in the log you've shared. It starts exactly like my one. The difference is that my one continues after that. See:
https://gist.github.com/janvorli/fbfafc88df226f604d76106f741b4320

**Comment by qmfrederik** [comment]
@cydhaselton I had a quick look, I was able to modify the CMake files enough so that it seems the coreclr build would pick up the tools in my NDK toolchain (created with `make-standalone-toolchain.sh `)

Have a look at [this commit](https://github.com/qmfrederik/coreclr/commit/d24c6367b47e4181679cd1f297781bba71f6097e) (or just checkout the features/android branch from my repo),

You would run it like `NDK_TOOLCHAIN_DIR=[path to your toolchain folder] ./build.sh arm64 cross`.

The output looks like this for me:

```
Commencing CoreCLR Repo build
WARNING: Can not determine runtime id for current distro.
Setting up directories for build
Checking prerequisites...
Laying out dynamically generated files consumed by the build system 
Laying out dynamically generated Event Logging Test files
Generating Event Logging Tests
Laying out dynamically generated Event Logging Implementation of Lttng
Cleaning the temp folder of dynamically generated Event Logging files
Updating /home/fcarlier/git/coreclr/bin/obj/Linux.arm64.Debug/Generated/eventprovider with /home/fcarlier/git/coreclr/bin/obj/Linux.arm64.Debug/Generated/eventprovider_new
Updating /home/fcarlier/git/coreclr/bin/obj/Linux.arm64.Debug/Generated/eventprovider/lttng
Updating /home/fcarlier/git/coreclr/bin/obj/Linux.arm64.Debug/Generated/eventprovider/tests
Updating /home/fcarlier/git/coreclr/bin/obj/Linux.arm64.Debug/Generated/eventprovider/CMakeLists.txt
Updating /home/fcarlier/git/coreclr/bin/obj/Linux.arm64.Debug/Generated/eventprovider/tracepointprovider
Commencing build of CoreCLR component for Linux.arm64.Debug in /home/fcarlier/git/coreclr/bin/obj/Linux.arm64.Debug
/home/fcarlier/git/coreclr
Running init-tools.sh
Tools are already initialized
Running: /home/fcarlier/git/coreclr/Tools/dotnetcli/dotnet /home/fcarlier/git/coreclr/Tools/run.exe /home/fcarlier/git/coreclr/config.json build -Project=/home/fcarlier/git/coreclr/build.proj -generateHeaderUnix -NativeVersionSourceFile=/home/fcarlier/git/coreclr/bin/obj/Linux.arm64.Debug/version.cpp -BuildArch=arm64 -BuildType=Debug -BuildOS=Linux
Running: /home/fcarlier/git/coreclr/Tools/msbuild.sh /nologo /verbosity:minimal /clp:Summary /maxcpucount  /l:BinClashLogger,Tools/net45/Microsoft.DotNet.Build.Tasks.dll;LogFile=binclash.log  /home/fcarlier/git/coreclr/build.proj /p:__BuildType=Debug /p:__BuildArch=arm64 /p:__BuildOS=Linux  /t:GenerateVersionSourceFile /p:GenerateVersionSourceFile=true /p:NativeVersionSourceFile=/home/fcarlier/git/coreclr/bin/obj/Linux.arm64.Debug/version.cpp       
Command execution succeeded.
Command successfully completed.
~/git/coreclr/bin/obj/Linux.arm64.Debug ~/git/coreclr
Invoking "/home/fcarlier/git/coreclr/src/pal/tools/gen-buildsys-clang.sh" "/home/fcarlier/git/coreclr" 3 6 arm64 Debug  Include_Tests  -DCLR_CMAKE_TARGET_OS=Linux -DCLR_CMAKE_PACKAGES_DIR=/home/fcarlier/git/coreclr/packages -DCLR_CMAKE_PGO_INSTRUMENT=0 
Including tests directory in build.
loading initial cache file /home/fcarlier/git/coreclr/cross/arm64/tryrun.cmake
-- The C compiler identification is GNU 4.9.0
-- The CXX compiler identification is GNU 4.9.0
-- Check for working C compiler: /home/fcarlier/toolchain//bin/aarch64-linux-android-gcc
-- Check for working C compiler: /home/fcarlier/toolchain//bin/aarch64-linux-android-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /home/fcarlier/toolchain//bin/aarch64-linux-android-g++
-- Check for working CXX compiler: /home/fcarlier/toolchain//bin/aarch64-linux-android-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The ASM compiler identification is GNU
-- Found assembler: /home/fcarlier/toolchain//bin/aarch64-linux-android-as
Detected Linux ARM64
CMake Error at src/corefx/System.Globalization.Native/CMakeLists.txt:13 (message):
  Cannont find utypes.h, try installing libicu-dev (or the appropriate
  package for your platform)


-- Configuring incomplete, errors occurred!
See also "/home/fcarlier/git/coreclr/bin/obj/Linux.arm64.Debug/CMakeFiles/CMakeOutput.log".
``` 

which matches what @janvorli was reporting on.

**Comment by janvorli** [comment]
@qmfrederik that's great! I just wonder why was the change in the root CMakeLists.txt needed - the android tools contain objcopy. Hmm, maybe just changing the line 
`find_program(OBJCOPY ${TOOLCHAIN}-objcopy)`
to
`find_program(OBJCOPY ${TOOLCHAIN_PREFIX}objcopy)`

would work?

**Comment by janvorli** [comment]
@qmfrederik I've just tried my suggestion and it works (and it would also work for the Linux crossbuild case)

**Comment by janvorli** [comment]
Hmm, there is a problem though. Now we would be building with gcc, which would not work for CoreCLR. We need to build with clang

**Comment by qmfrederik** [comment]
@janvorli Yup, tried the same thing, works here as well - I've pused a [rebased commit](https://github.com/qmfrederik/coreclr/commit/d51b7d090cfc07a2e0f8a91f2d56ca44025d4828)

**Comment by qmfrederik** [comment]
Well, I guess we can just update the `CMAKE_CXX_COMPILER` flags to point to clang; the NDK includes clang. But it's getting late in my timezone :)

**Comment by janvorli** [comment]
Updating CMAKE_CXX_COMPILER and CMAKE_C_COMPILER to point to the clang in the NDK worked:
set(CMAKE_C_COMPILER ${CROSS_NDK_TOOLCHAIN}/bin/clang)
set(CMAKE_CXX_COMPILER ${CROSS_NDK_TOOLCHAIN}/bin/clang++)

**Comment by cydhaselton** [comment]
>  I don't see any error in the log you've shared. It starts exactly like my one. The difference is that my one continues after that. See:

The log I posted was truncated.  I'll try again with the changes and if it fails I'll post the full log.

**Comment by cydhaselton** [comment]
Configureonly ran with the changes until it failed on a missing libunwind, which I can install for the NDK.

**Comment by cydhaselton** [comment]
Currently digging into what appears to be a pthread issue (CMakeError.log here: http://pastebin.com/K3DiK18U). Android has pthread integrated into libc; libpthread doesn;t exist.

**Comment by janvorli** [comment]
I don't think this is a problem. This just detects if these features are present, CoreCLR can build without them. Moreover, we don't use them on ARM / ARM64 anyways:
```C++
// Temporarily disabling usage of pthread process-shared mutexes on ARM/ARM64 due to functional issues that cannot easily be
// detected with code due to hangs. See https://github.com/dotnet/coreclr/issues/5456.
#if HAVE_FULLY_FEATURED_PTHREAD_MUTEXES && HAVE_FUNCTIONAL_PTHREAD_ROBUST_MUTEXES && !(defined(_ARM_) || defined(_ARM64_))
    #define NAMED_MUTEX_USE_PTHREAD_MUTEX 1
#else
    #define NAMED_MUTEX_USE_PTHREAD_MUTEX 0
#endif
```

**Comment by cydhaselton** [comment]
@janvorli I;m going by the last few lines in the CMakeError.log; the build seems to fail at detecting pthread features. I can add `DHAVE_FULLY_FEATURED_PTHREAD_MUTEXES` to cmakeargs to see if the build progresses

**Comment by janvorli** [comment]
This is not a build error during the build, this is a build error while detecting a feature, which is perfectly fine.

**Comment by janvorli** [comment]
What error do you see in the console when you run the build.sh command?

**Comment by cydhaselton** [comment]
The configuration fails at that point; the detection seems to be picking up something that prevents the buikd from progressing.

I'll upload the latest CMakeError.log; the last error is `cannot find -lpthread`

**Comment by janvorli** [comment]
@cydhaselton the CMakeError.log contains a lot of errors, that's natural, since the first step of build is to detect which stuff is present on the build machine, which functions are not there etc. It does it by attempting to compile simple pieces of code and when something is not there, it just reports compiler errors to this log, but it goes on. That's why I have asked for the log from the console instead, which is much more important to see what is the real problem. 
If you can see in the console that it complains about the -lpthread, then the problem is that you need to remove the reference to it from CMakeLists.txt files conditionaly for Android. This will be tedious since every PAL test contains something like:
```
target_link_libraries(paltest_waitforsingleobject_wfsomutextest
  pthread
  m
  coreclrpal
)
```
To make the pthread conditional, you'll need to e.g. create a variable PTHREAD, replace all usages of pthread in target_link_libraries by ${PTHREAD} and set it to "pthread" only for non-android builds.

**Comment by janvorli** [comment]
Or, maybe there would be a way to create a dummy library named pthread on Android and then you'd not need to modify the CMAkeLists.txt, which would be preferrable. I don't mean creating a real library, but rather using CMake means. Let me see if / how it can be done.

**Comment by janvorli** [comment]
Hmm, it seems that I cannot do it in CMake. But you can do a quick hack. Create an empty dummy.cpp file in the root folder and add this line to the root CMakeLists.txt e.g. at line 541 (after the `include(pgosupport.cmake)`)
```cmake
add_library(pthread STATIC dummy.cpp)
```
This should fix the problem for now, we can do the correct final solution later.

**Comment by cydhaselton** [comment]
@janvorli The configure fails in the same location, even with the dummy.cpp file and the added line in CMakeFiles.txt. The console output is here: http://pastebin.com/8WvXY2bR, but there doesn;t seem to be a lot of useful info, except for where it stopped.

Note: if I run ./build.sh with skiptests and skipnative, configure and build runs successfully.

**Comment by janvorli** [comment]
@cydhaselton, this is the problem:
```
CMake Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:
   HAVE_COMPATIBLE_EXP_EXITCODE (advanced)
```
What is suggests is to add the HAVE_COMPATIBLE_EXP_EXITCODE to the cache in the cross/arm64/tryrun.cmake with value reflecting what the result of the test would be on the real target.
So you need to add
```cmake
SET( HAVE_COMPATIBLE_EXP_EXITCODE 
     0
     CACHE STRING "Result from TRY_RUN" FORCE)
```
The value would be 0 or 1, I have no idea what would be the value it would get when running on Android.
If you can run the following over there and check the return value, you can use it:
```C++
#include <math.h>
#include <stdlib.h>

int main(void) {
  double d = exp(1.0), e = M_E;

  /* Used memcmp rather than == to test that the doubles are equal to
   prevent gcc's optimizer from using its 80 bit internal long
   doubles. If you use ==, then on BSD you get a false negative since
   exp(1.0) == M_E to 64 bits, but not 80.
  */

  if (memcmp (&d, &e, sizeof (double)) == 0) {
    exit(0);
  }
  exit(1);
}
```

**Comment by cydhaselton** [comment]
@janvorli Had to add a `printf` to see the output value but otherwise running the above on Android worked.

Running into an error consistently during the part of the build when verify-so.sh is run.  I noticed that there's an `ifthen` that prevents the check from being run on Alpine Linux; should Android be added to this check as well?

**Comment by janvorli** [comment]
@cydhaselton yes, that verification is meant to be executed on Linux only, so for android, it needs to be skipped.

**Comment by cydhaselton** [comment]
@janvorli is there documentation on where the `$ID` variable is generated?  Originally I thought it was via setting `-officialbuildid ` when running `./build.sh` but grepping for officialbuildid shows it's referenced only in `./build.sh` and nowhere else.

**Comment by janvorli** [comment]
The $ID comes from /etc/os-release (imported via the source $ROOTFS_DIR/etc/os-release). I guess Android doesn't have it, right? I would suggest forging something based on the /etc/os-release from your Linux box.

**Comment by qmfrederik** [comment]
@cydhaselton Out of curiosity, how did you resolve the libicu dependency? And what value did you get for the `HAVE_COMPATIBLE_EXP_EXITCODE` variable?

**Comment by cydhaselton** [comment]
@janvorli: Couple of options for this, depending on what the `function(verify_dependencies targetName errorMessage)` in `functions.cmake` should return or if it should run at all

1. Create an alternate /etc/os-release for the build with appropriate entries for the target platform (`cp /etc/os-release /etc/cross-release; nano /etc/cross-release`) and modify `verify-so.sh`…something like
```
if [ "$ID" != "alpine" ]; then
   source /etc/cross-release
   if ["$ID" != "android" ]; then
   ...
```
2. Set an environment variable like `export CROSS=ANDROID` and add a detection for it after line 223 in `functions.cmake`. Something like 
```
# We don't need to verify dependencies on OSX, since missing dependencies
    # result in link error over there.
    if (NOT CLR_CMAKE_PLATFORM_DARWIN)
    if (NOT $ENV{CROSS})
...
```

If `function(verify_dependencies targetName errorMessage)` doesn;t need to run at all, the detection can happen before.

**Comment by cydhaselton** [comment]
> @qmfrederik: Out of curiosity, how did you resolve the libicu dependency? And what value did you get for the `HAVE_COMPATIBLE_EXP_EXITCODE` variable?

The Termux app on device has a libicu package available; it can be installed and the appropriate libs/headers copied to the ndk sysroot. If Termux is not available, cross building libicu for Android is trivial.

Running the code @janvorli provided resulted in a value of `1` for `HAVE_COMPATIBLE_EXP_EXITCODE`

**Comment by qmfrederik** [comment]
@cydhaselton Cool, thanks! 

I had a quick look around; the termux packages are available here: [libicu-dev_58.2_aarch64.deb](http://termux.net/dists/stable/main/binary-aarch64/libicu-dev_58.2_aarch64.deb), [libicu_58.2_aarch64.deb](http://termux.net/dists/stable/main/binary-aarch64/libicu_58.2_aarch64.deb); you can find the other packages at http://termux.net/dists/stable/main/binary-aarch64/ and the other architectures at http://termux.net/dists/stable/main/.

So if there are other packages that turn out to be missing, or you want to build for another Android architecture, you should be able to just download the packages from there and use them (`.deb` files are just compressed packages).

**Comment by qmfrederik** [comment]
@cydhaselton cool, that actually worked for me - the libicu-related headers are now found correctly; I've shared the [Dockerfile](https://github.com/qmfrederik/dotnet-alpine-arm/blob/master/android/Dockerfile) I use for building.

The next step seems to be libunwind. Did you see a missing libunwind dependency? Did you find a prepackaged version of libunwind or did you have to compile it yourself? (Just trying to avoid double work :))

Note: Android seems to ship with libunwind OOB, let me figure out why I'm getting that error.

**Comment by qmfrederik** [comment]
hmm, I can get libunwind to cross compile using, roughly, these commands:

```
git clone https://android.googlesource.com/platform/external/libunwind/ /home/libunwind
cd /home/libunwind
autoreconf --force -v --install
./configure CC=aarch64-linux-android-gcc CPPFLAGS="-I/home/toolchain/include/" CFLAGS="-nostdlib" LDFLAGS="-Wl,-rpath-link=/home/toolchain/lib/ -L/home/toolchain/lib/" LIBS="-lc -lgcc" --host=x86_64 --target=aarch64-eabi --disable-coredump --prefix=/home/toolchain/
make
make install
```

yet the CMakeFiles don't pick up the libunwind-aarch64.h header; let me figure out what's wrong there.

**Comment by cydhaselton** [comment]
@qmfrederik My build fails a bit earlier, when running the `verify-so.sh` script from `functions.cmake`. I've written a few workarounds but I'm not sure if the verify-so.sh needs to return in order for the build to progress or if it can be skipped entirely (see previous post); the build errors out immediately after linking System.Globalization.Native.so.

I built libunwind in Termux so I just copied the headers/libs to the ndk sysroot. I had problems with debugging in gdb that were either due to the Termux gdb or the libunwind I built so I may run into issues again once I get further in the build.

**Comment by qmfrederik** [comment]
@cydhaselton It's strange, I never had to change the `verify-so.sh` script but I think I'm at pretty much the same stage you are at - the configuration phase passes, the actual build starts, and I'm seeing  a failure at about 1% - the last log entry is linking `System.Globalization.Native.so`.

Because the build script runs multiple tasks in parallel, one task may fail while others still continue, so the percentages are not always accurate and the actual errors may have been printed out a log earlier in the build process.

In my case, I'm seeing the following failure:

```
/home/coreclr/src/pal/src/include/pal/threadsusp.hpp:77:2: error: "Don't know how to synchronize thread suspends and resumes on this platform"
#error "Don't know how to synchronize thread suspends and resumes on this platform"
 ^
```

I'll have a look; if you've already worked around it let me know.

**Comment by qmfrederik** [comment]
It looks like you can get around the first error by changing line 77 in `src/pal/src/include/pal/threadsusp.hpp:77` to `#include <semaphore.h>`.

That leaves this:
```
/home/coreclr/src/pal/src/cruntime/file.cpp[  0%] Building CXX object tests/src/JIT/Regression/VS-ia64-JIT/V1.2-M02/b108129/CMakeFiles/test2.dir/test2.cpp.o
:228:11: error: use of undeclared identifier 'getw'
    ret = getw( f->bsdFilePtr );
          ^
/home/coreclr/src/pal/src/cruntime/file.cpp:481:11: error: use of undeclared identifier 'putw'
    ret = putw(c,  f->bsdFilePtr );
          ^
```

**Comment by cydhaselton** [comment]
@qmfrederik If you're building for alpine, verify-so.sh is not run; there's a check for it in the script.

I'll try the line 77 change. Meanwhile check the `file.cpp` in the feature-android branch of my fork; I believe it has Android-specific edits

**Comment by qmfrederik** [comment]
@cydhaselton My messages in this thread relate to my attempts to cross-build .NET Core for Android on an Ubuntu host; yet I didn't have to make changes to `verify-so.sh`. Strange.

I currently can get building up to ~16%, at which point the build fails like this:

```
/home/toolchain/bin/aarch64-linux-android-as: unrecognized option '-ARM64'
src/vm/wks/CMakeFiles/cee_wks.dir/build.make:4718: recipe for target 'src/vm/wks/CMakeFiles/cee_wks.dir/__/arm64/asmhelpers.S.o' failed
```

The full command which is being executed is:

```
/home/toolchain/bin/aarch64-linux-android-as -DARM64 -DBIT64=1 -DBUILDENV_CHECKED=1 -DDBG_TARGET_64BIT=1 -DDBG_TARGET_ARM64=1 -DDBG_TARGET_ARM64_UNIX -DDBG_TARGET_WIN64=1 -DDEBUG -DDEBUGGING_SUPPORTED -DDISABLE_CONTRACTS -DFEATURE_APPDOMAIN_RESOURCE_MONITORING -DFEATURE_ARRAYSTUB_AS_IL -DFEATURE_ASYNC_IO -DFEATURE_COLLECTIBLE_TYPES -DFEATURE_CORECLR -DFEATURE_COREFX_GLOBALIZATION -DFEATURE_CORESYSTEM -DFEATURE_CORRUPTING_EXCEPTIONS -DFEATURE_DBGIPC_TRANSPORT_DI -DFEATURE_DBGIPC_TRANSPORT_VM -DFEATURE_EVENTSOURCE_XPLAT=1 -DFEATURE_EXCEPTIONDISPATCHINFO -DFEATURE_EXCEPTION_NOTIFICATIONS -DFEATURE_HIJACK -DFEATURE_HOST_ASSEMBLY_RESOLVER -DFEATURE_ICASTABLE -DFEATURE_IMPLICIT_TLS -DFEATURE_ISYM_READER -DFEATURE_LEAVE_RUNTIME_HOLDER=1 -DFEATURE_LEGACYNETCF_DBG_HOST_CONTROL -DFEATURE_LOADER_OPTIMIZATION -DFEATURE_MAIN_CLR_MODULE_USES_CORE_NAME -DFEATURE_MANAGED_ETW_CHANNELS -DFEATURE_MERGE_CULTURE_SUPPORT_AND_ENGINE -DFEATURE_MULTICOREJIT -DFEATURE_MULTIREG_RETURN -DFEATURE_NORM_IDNA_ONLY -DFEATURE_PAL -DFEATURE_PAL_ANSI -DFEATURE_PAL_SXS -DFEATURE_PERFMAP -DFEATURE_PREJIT -DFEATURE_RANDOMIZED_STRING_HASHING -DFEATURE_READYTORUN -DFEATURE_SPAN_OF_T -DFEATURE_STANDALONE_SN -DFEATURE_STRONGNAME_DELAY_SIGNING_ALLOWED -DFEATURE_STRONGNAME_MIGRATION -DFEATURE_STUBS_AS_IL -DFEATURE_SVR_GC -DFEATURE_SYMDIFF -DFEATURE_SYNCHRONIZATIONCONTEXT_WAIT -DFEATURE_SYNTHETIC_CULTURES -DFEATURE_USE_ASM_GC_WRITE_BARRIERS -DFEATURE_VERSIONING -DFEATURE_WINDOWSPHONE -DFEATURE_WINMD_RESILIENT -DLINUX64 -DPLATFORM_UNIX=1 -DPROFILING_SUPPORTED -DUNICODE -DURTBLDENV_FRIENDLY=Checked -DWRITE_BARRIER_CHECK=1 -D_ARM64_ -D_BLD_CLR -D_DBG -D_DEBUG -D_SECURE_SCL=0 -D_TARGET_ARM64_=1 -D_UNICODE -D_WIN64 -I/home/coreclr/bin/obj/Linux.arm64.Debug/src/vm/wks -I/home/coreclr/src/vm/wks -I/home/coreclr/src/vm -I/home/coreclr/src/pal/prebuilt/inc -I/home/coreclr/bin/obj -I/home/coreclr/src/pal/inc -I/home/coreclr/src/pal/inc/rt -I/home/coreclr/src/pal/src/safecrt -I/home/coreclr/src/inc -I/home/coreclr/src/strongname/inc -I/home/coreclr/src/inc/winrt -I/home/coreclr/src/debug/inc -I/home/coreclr/src/debug/inc/arm64 -I/home/coreclr/src/debug/inc/dump -I/home/coreclr/src/md/inc -I/home/coreclr/src/classlibnative/bcltype -I/home/coreclr/src/classlibnative/cryptography -I/home/coreclr/src/classlibnative/inc -I/home/coreclr/bin/obj/Linux.arm64.Debug/src/inc -I/home/coreclr/src/pal/inc/rt/cpp -I/home/coreclr/src/nativeresources -I/home/coreclr/src/gc -I/home/coreclr/src/vm/arm64 -Wa,--noexecstack -g --sysroot=/home/toolchain/sysroot -fno-omit-frame-pointer -fms-extensions -fwrapv -fstack-protector-strong -ferror-limit=4096 -Werror -Wno-unused-private-field -Wno-unused-variable -Wno-microsoft -Wno-tautological-compare -Wno-constant-logical-operand -Wno-unknown-warning-option -Wno-invalid-offsetof -Wno-incompatible-ms-struct -fsigned-char -nostdinc -fPIC -o CMakeFiles/cee_wks.dir/__/arm64/asmhelpers.S.o -c /home/coreclr/src/vm/arm64/asmhelpers.S
```

and it looks like the `-DARM64` flag is being coverted to `-ARM64`; if I remove that one it just complains about `BIT64` (the second argument in the list)

My entire build script is in this [Dockerfile](https://github.com/qmfrederik/dotnet-alpine-arm/blob/master/android/Dockerfile) + the repo which is referenced from there.

Did you get an error like this?

**Comment by qmfrederik** [comment]
... it looks like somehow flags for the C compiler are being passed to as, really strange.

**Comment by qmfrederik** [comment]
Oh, so it's the other way round - it appears (to be confirmed) the build scripts expect to use clang (?) to compile `.S` files, so I've updated `cross/arm64/toolchain.cmake` to account for that.

The build now compiles up to 36% and fails with this error:

```
[ 36%] Linking CXX shared library libclrjit.so
src/md/compiler/crossgen/CMakeFiles/mdcompiler_crossgen.dir/__/regmeta_vm.cpp.o
/home/toolchain/bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find -lgcc_s
/home/toolchain/bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find -lpthread
/home/toolchain/bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find -lrt
clang38++: error: linker command failed with exit code 1 (use -v to see invocation)
```

so let's find out what's up with those files on Android.

**Comment by cydhaselton** [comment]
@qmfrederik The next errors I'm getting are during the asm build; `aarch64-linux-android-as` does not recognize at least two of the defines, `-DARM64`, `-BIT64=1`, and probably more:

```
/ndk/bin/aarch64-linux-android-as: unrecognized option '-BIT64=1'
make[2]: *** [src/vm/wks/CMakeFiles/cee_wks.dir/__/arm64/asmhelpers.S.o] Error 1
```

What changes did you make to `toolchain.cmake` to correct this?

**Comment by cydhaselton** [comment]
@qmfrederik There is no `-lpthread` on Android; thread functions are built in to Bionic. `-lgcc` should be in the ndk; not sure about `-lrt`

EDIT: looks like -lrt functionality is also integrated into Bionic. There's a suggestion to create symlinks to libc called librt and libpthread respectively

**Comment by qmfrederik** [comment]
@cydhaselton It looks like we're almost there - the build continues up to 100% on my side, but fails at the very last moment:

```
In file included from /home/coreclr/src/ToolBox/SOS/Strike/sos.h:7:
In file included from /home/coreclr/src/ToolBox/SOS/Strike/strike.h:57:
/home/toolchain/sysroot/usr/include/wchar.h:46:18: error: expected member name or ';' after declaration specifiers
  char __reserved[4];
  ~~~~           ^
```

I'll have a look at it tomorrow if time permits.

**Comment by cydhaselton** [comment]
The next build error I see is about 40% in

```
/usr/bin/ranlib libcee_crossgen.a                                                          
 make[2]: Leaving directory `/build/core/coreclr/bin/obj/Linux.arm64.Debug'                  /usr/bin/cmake -E cmake_progress_report /build/core/coreclr/bin/obj/Linux.arm64.Debug/CMakeFiles  2 3 4 5                                                                               [ 40%] Built target cee_crossgen                                                            make[1]: Leaving directory `/build/core/coreclr/bin/obj/Linux.arm64.Debug'                  
make: *** [all] Error 2                                                                     
Failed to build coreclr components.
```

Not sure if this is due to the host ranlib being used on `libcee_crossgen.a` or not; investigating

**Comment by cydhaselton** [comment]
@qmfrederik How are you getting the build close to 100%?
Additionally, I have to run the `clean.sh` script before I rebuild after an error; otherwise I get a slew of errors (see log: http://pastebin.com/XvD9tnE9). Do you run into the same issue?

**Comment by cydhaselton** [comment]
> @qmfrederik:  It looks like we're almost there - the build continues up to 100% on my side, but fails at the very last moment:

`__reserved` needs to be renamed, it conflicts with a definition in Android; see the `feature-android` branch

Also please update with changes you made to `toolchain.cmake` and others that allowed build to progress

**Comment by qmfrederik** [comment]
@cydhaselton Apologies, somehow I don't see new notifications in real time in the browser, so I missed some of your messages.

- In your toolchain.cmake, set `CMAKE_ASM_COMPILER` to `clang`, see my [toolchain.cmake](https://github.com/qmfrederik/coreclr/blob/features/android/cross/arm64/toolchain.cmake#L14)
- As to `pthread` and the other libraries, I used @janvorli's trick to create a dummy library and that worked fine - no symlinks required. I had to add references to the libraries I re-used from Termux; see [src/pal/src/CMakeLists.txt](https://github.com/qmfrederik/coreclr/blob/features/android/src/pal/src/CMakeLists.txt#L300)
- I don't see the errors you get when rebuilding, but I noticed you branched off an older version of coreclr than I did. I occasionally do `rm -rf bin` or `git clean -xfd` just to make sure CMake didn't cache anything, but that's about it.
- The build output you copied doesn't contain an error, the error should be earlier in your build output (usually the line with the error contains the text `error:`). Because the build runs in parallel, one task may error out and others may still continue, so the error message is not necessarily the last one. If you want to avoid that, cd into `bin/obj/Linux.arm64.Debug/` and run `make -j 1`, that way your build process will be single-threaded (and take more time).

Would be cool if the `__reserved` issue can be solved, that should get us a compiling version of CoreCLR.

**Comment by qmfrederik** [comment]
My entire build process is on GitHub, see this [Dockerfile](https://github.com/qmfrederik/dotnet-alpine-arm/blob/master/android/Dockerfile) and my [features/android](https://github.com/qmfrederik/coreclr/tree/features/android) branch.

If you want to build the exact same way I did:

- Install Docker (`apt-get install docker.io` if you're on Ubuntu)
- Clone the https://github.com/qmfrederik/dotnet-alpine-arm/
- In the `android` directory of that repo, run: `docker build -t dotnetandroid . && docker run --name dotnetandroid -ti dotnetandroid`
- That will get you a bash prompt running inside that container.
- Then, run `ROOTFS_DIR=${NDK_SYSROOT_DIR} ./build.sh cross arm64 skipgenerateversion skipmscorlib cmakeargs -DENABLE_LLDBPLUGIN=0` to kick off the build.

**Comment by cydhaselton** [comment]
@qmfrederik: Couple of items:

- There are patches for the  __reserved issue in the `feature-android` branch of this fork; let me know if you can;t find them and I'll post them here. 
- What;s the benefit of running the cross build in Docker/alpine as opposed to running it in Ubuntu 14.04?
- The re-build errors are here: http://pastebin.com/XvD9tnE9. This happens if I re-run `./build.sh params` without running `clean.sh` first.

**Comment by qmfrederik** [comment]
@cydhaselton 

As to the docker container; the repo name is a bit misleading. The OS running in the container is Ubuntu, not Alpine. I prefer using docker containers (or Vagrant VMs) when working on build processes. They make it easy to test the entire build process on a vanilla configuration. You can easily destroy & rebuild containers, forcing you to script (and document) the entire build process.

I've cherry-picked https://github.com/cydhaselton/coreclr/commit/54156b230079b87e222724a73ea9bad34f839d4d and CoreCLR now cross-compiles for Android!

I've also rebased my branch against upstream/master, as to start from a more or less clean slate.

I guess the next step is to run the PAL tests on an Android machine. I would like to avoid mixing the aarch64 bring-up and Android bring-up in one single task, so I'll probably attempt to build CoreCLR for x64 Android and armhf Android.

I'll also take some time to review my git history and rewrite it in a way that would prepare the individual commits for upstream PRs.

I'll keep you posted - let me know how things go on your end!

**Comment by cydhaselton** [comment]
@qmfrederik Latest issue: for some reason the build isn't finding liblzma, even though it is in /ndk/sysroot/usr/lib (log: http://pastebin.com/t64ycfJd). Setting -v doesn't reveal more than the error.

Digging into build dir, let me know if you ran into this issue

**Comment by cydhaselton** [comment]
@qmfrederik Still working on the lzma issue…did you run into this? Also, in order to do much of anything, corefx needs to be built as well; were you able to do that?

**Comment by qmfrederik** [comment]
@cydhaselton I'm sorry, I didn't see this issue. Which `lzma`-related files do you have in `/ndk/sysroot/usr/lib`?

**Comment by cydhaselton** [comment]
@qmfrederik The lmza library files…`liblzma.so`. The file is there but for whatever reason the linker doesn't see them. I'm wondering if it has something to do with the build using `clang38++` instead of `aarch64-linux-android-clang++`

Also I'm getting an error with your Docker image; at the very end it displays the following:
```
inflating: android-ndk-r13b/ndk-build                                                     
INFO[1462] ApplyLayer exit status 1 stdout:  stderr: write /opt/android-ndk-tmp/android-ndk-r13b/platforms/android-21/arch-mips/usr/libr6/libm.a: read-only file system
```

and doesn't start the Docker image prompt

**Comment by cydhaselton** [comment]
@qmfrederik Re-running the Docker command results in this:
```
INFO[0000] Error running DeviceCreate (createSnapDevice) dm_task_run failed
```

**Comment by qmfrederik** [comment]
@cydhaselton Do you have enough free disk space? What does `df -lh` give you?

**Comment by qmfrederik** [comment]
@cydhaselton I've just pushed some scripts which should allow you to cross-compile CoreCLR for Android to my repo:

https://github.com/qmfrederik/coreclr/commit/efbbb7a8624d624c37a1673d2686d92158414558

The script will download the NDK, set up the toolchain and sysroot including libunwind and any dependencies borrowed from Termux.

You should be able to either build from my branch, or cherry-pick my commit into your branch and build from there:

#### Building from my branch

```
git clone -b features/android https://github.com/qmfrederik/coreclr
cd coreclr
./init-android.sh
CONFIG_DIR=`realpath cross/android/arm64` ROOTFS_DIR=`realpath cross/android/toolchain/arm64/sysroot` ./build.sh cross arm64 skipgenerateversion skipmscorlib cmakeargs -DENABLE_LLDBPLUGIN=0
```


#### Cherry-picking my commit into your branch
```
cd coreclr
git remote add qmfrederik https://github.com/qmfrederik/coreclr
git fetch qmfrederik # This doesn't make any changes to your branch
git cherry-pick efbbb7a8624d624c37a1673d2686d92158414558 # This will apply this specific commit only to your branch
./init-android.sh
CONFIG_DIR=`realpath cross/android/arm64` ROOTFS_DIR=`realpath cross/android/toolchain/arm64/sysroot` ./build.sh cross arm64 skipgenerateversion skipmscorlib cmakeargs -DENABLE_LLDBPLUGIN=0
```

Let me know how it goes.

**Comment by cydhaselton** [comment]
@qmfrederik 
> Do you have enough free disk space? What does df -lh give you?

Definitely not enough; I'm working with a free AWS EC2 instance with a 15GB drive. How big is the docker image?

> The script will download the NDK, set up the toolchain and sysroot including libunwind and any dependencies borrowed from Termux.
> 
> You should be able to either build from my branch, or cherry-pick my commit into your branch and build from there:

Excellent! I'll give that a shot and post back with results.

Did you see my question about corefx?

**Comment by qmfrederik** [comment]
@cydhaselton Didn't try corefx yet; as a next step, you should already be able to run the PAL tests on Android :)

Wanted to check whether I could get coreclr to compile for x86 or x64 Android because it would enable working on emulators; but libunwind seems to get in the way.

Let me know if you have success with the PAL tests/corefx native build!

**Comment by cydhaselton** [comment]
@qmfrederik: There are some errors with the android build script; Is your goal to submit a pull request for `init-android.sh`?

**Comment by qmfrederik** [comment]
@cydhaselton Yup, if we can get the Android port more or less working, I think it would make sense to get the `init-android.sh` script in the upstream repo, just like there's a script to initialize the rootfs for cross-building for ARM Linux.

So if you find issues, please let me know. 

Did the cross-build of coreclr succeed on your side?

**Comment by cydhaselton** [comment]
@qmfrederik FYI slammed at work so I may be out of pocket for the next day or two.  I'll get the build log uploaded and a link posted as soon as I can.

The cross-build did not succeed on my side; there are breaking errors thrown by the `init-android.sh` script in your fork, I don't have enough space on the build server I'm using for your Docker image and I'm still getting the `cannot find -lzma` error when building in my feature-android branch (full log here:  http://pastebin.com/t64ycfJd)

**Comment by qmfrederik** [comment]
@cydhaselton No worries.

In meanwhile, I got the PAL tests to run. I compiled for arm64 with the build script; I needed to make one additional change so that the compile would compile PIE code.

Once that was done, I ran the following commands to copy the PAL tests over to an Android phone:

```
adb push bin/obj/Linux.arm64.Debug/src/pal/tests/palsuite/ /data/local/tmp/coreclr/src/pal/tests/palsuite
adb push cross/android/toolchain/arm64/sysroot/usr/lib/libuuid.so.1 /data/local/tmp/coreclr/lib
adb push cross/android/toolchain/arm64/sysroot/usr/lib/libintl.so /data/local/tmp/coreclr/lib
adb push cross/android/toolchain/arm64/sysroot/usr/lib/libandroid-support.so /data/local/tmp/coreclr/lib/
adb push cross/android/toolchain/arm64/sysroot/usr/lib/libandroid-glob.so /data/local/tmp/coreclr/lib/
adb push src/pal/tests/palsuite/paltestlist.txt /data/local/tmp/coreclr
adb push src/pal/tests/palsuite/runpaltests.sh /data/local/tmp/coreclr/
```

and, then I ran `adb shell` to launch a shell on Android. From there, I could launch the PAL tests:

```
./runpaltests.sh /data/local/tmp/coreclr/
```

with this output:

```
The following test(s) failed:
exception_handling/pal_sxs/test1/paltest_pal_sxs_test1. Exit code: 1
filemapping_memmgt/ProbeMemory/ProbeMemory_neg1/paltest_probememory_probememory_neg1. Exit code: 133
file_io/CopyFileA/test4/paltest_copyfilea_test4. Exit code: 1
file_io/CopyFileW/test3/paltest_copyfilew_test3. Exit code: 1
threading/NamedMutex/test1/paltest_namedmutex_test1. Exit code: 1

PAL Test Results:
  Passed: 785
  Failed: 5
```

**Comment by qmfrederik** [comment]
... and here's how to use LLDB to debug a PAL test:

- Download LLDB & unzip from https://dl.google.com/android/repository/lldb-2.3.3614996-linux-x86_64.zip
- Push the lldb server server to Android: `adb push lldb/2.2/android/arm64-v8a/lldb-server /data/local/tmp/`
- Launch the lldb server on the Android device, have it listen on port 1234 (or any other port): 
```
adb shell
cd /data/local/tmp
./lldb-server platform --listen *:1234
```
- Set up port forwarding from Android to your PC: `adb forward tcp:1234 tcp:1234`
- Install lldb 3.9 on your PC and connect to your Android device:
```
lldb-3.9
(lldb) platform select remote-android
  Platform: remote-android
 Connected: no
(lldb) platform connect connect://localhost:1234
  Platform: remote-android
    Triple: aarch64-*-linux-android
OS Version: 23.0.0 (3.10.84-perf-gf38969a)
    Kernel: #1 SMP PREEMPT Fri Sep 16 11:29:29 2016
  Hostname: localhost
 Connected: yes
WorkingDir: /data/local/tmp

(lldb) target create coreclr/src/pal/tests/palsuite/file_io/CopyFileA/test4/paltest_copyfilea_test4
(lldb) env LD_LIBRARY_PATH=/data/local/tmp/coreclr/lib
(lldb) run
Process 26040 launched: '/home/fcarlier/.lldb/module_cache/remote-android/.cache/8B120569-9FDE-9028-C6C4-C4729F516754-BB027E8B/paltest_copyfilea_test4' (aarch64)
CopyFileA: Cannot copy a file with error, 2
Process 26040 exited with status = 1 (0x00000001) 
```

**Comment by qmfrederik** [comment]
ok - guessing the pal tests fail on Android because they interact with `/etc/passwd`, a file which doesn't exit on Android. So the PAL tests look OK.

**Comment by cydhaselton** [comment]
@qmfrederik Here's the log of the build script errors. Warning: it's large so you probably want to wget it and grep locally
http://pastebin.com/QuTMgW6n

**Comment by qmfrederik** [comment]
@cydhaselton I had a quick look at the log but didn't see any errors.

Did you run
```
CONFIG_DIR=`realpath cross/android/arm64` ROOTFS_DIR=`realpath cross/android/toolchain/arm64/sysroot` ./build.sh cross arm64 skipgenerateversion skipmscorlib cmakeargs -DENABLE_LLDBPLUGIN=0
```

as instructed at the end of the log?

**Comment by qmfrederik** [comment]
@cydhaselton I'm giving CoreFX a try; the first dependency I fail to find "out of the box" is krb5. Did you cross compile it from scratch or did you find a packaged version somewhere?

I see there's a Termux PR (termux/termux-packages#588) open for it but it's not there yet.

**Comment by cydhaselton** [comment]
@qmfrederik I compiled it from scratch in Termux.

I didn't run the command at the end of the log because a number of items failed during the build.

1. All of the downloaded termux packages fail to unpack:
```
Unpacking Termux packages
dpkg-deb: error: archive '/build/core2/coreclr/cross/android/deb/libicu_58.2_x86_64.deb' has premature member 'control.tar.xz' before 'control.tar.gz', giving up
dpkg-deb: error: archive '/build/core2/coreclr/cross/android/deb/libicu-dev_58.2_x86_64.deb' has premature member 'control.tar.xz' before 'control.tar.gz', giving up
dpkg-deb: error: archive '/build/core2/coreclr/cross/android/deb/libuuid-dev_1.0.3_x86_64.deb' has premature member 'control.tar.xz' before 'control.tar.gz', giving up
[snip]
```
2. Downloads for the libuuid packages fail:
```
http://termux.net/dists/stable/main/binary-i686/libuuid-dev_1.0.3_i686.deb:
2017-01-24 02:17:23 ERROR 404: Not Found.
http://termux.net/dists/stable/main/binary-i686/libuuid_1.0.3_i686.deb:
2017-01-24 02:17:23 ERROR 404: Not Found.
```
3. Unknown toolchain API (i.e. android-xx, indicating the target version of Android...without knowing this the build could succeed but the generated binaries fail to run on Android as there were significant changes in requirements between JellyBean and KitKat, KitKat and Lollipop)
```
Generating the arm64 toolchain
Generating the x64 toolchain
Generating the x86 toolchain
```

The build command would've failed due to the missing Termux packages.  Also, need to know what platform the toolchain was built for.

Re-check the posted build log...I pulled the above info from there.

**Comment by qmfrederik** [comment]
@cydhaselton Ah, sorry for that! I actually did some work on that branch today and split off a new one - features/android-rebased . It cleans up a lot of stuff - for now, only arm64 is being targetted.

Still, except for 3., the errors you listed shouldn't preven tyou from cross-compiling for arm64. What's the unknown toolchain API error you are referring to? Do you have a more specific error message?

**Comment by cydhaselton** [comment]
@qmfrederik It's not an error message.  The `make-standalone-toolchain.sh` in the NDK can target a variety of Android platforms; you have to choose which one to target.  Did you read my previous message?

Regarding the errors, how would coreclr build for Android/arm64 without the prerequisites downloaded from Termux?

**Comment by qmfrederik** [comment]
@cydhaselton Android is backward compatible, API level 23 matches Android 6.0 so code compiled with API level 23 will work with 6.0 and above (and most likely with earlier versions as well, unless we happen to rely on a function which was not available in an earlier version).

Looking at https://developer.android.com/about/dashboards/index.html, it may make sense to target, say Android 4.4 instead (API level 19). You should be able to update the script and rebuild. If you compile for 4.4, it will work with all later versions.

Regarding the errors, the script downloads dependencies for four architectures:
- arm64 (aarch64)
- arm
- x64 (x86_64)
- x86 (i686)

The errors are for i686; so the toolchain for arm64 is still OK.

That said, it's confusing, and it has been fixed on the features/android-rebased branch.

**Comment by cydhaselton** [comment]
> Android is backward compatible, API level 23 matches Android 6.0 so code compiled with API level 23 will work with 6.0 and above (and most likely with earlier versions as well, unless we happen to rely on a function which was not available in an earlier version).

@qmfrederik Android has been known to remove functions between API levels; I've run into this situation before.   

> Looking at https://developer.android.com/about/dashboards/index.html, it may make sense to target, say Android 4.4 instead (API level 19). You should be able to update the script and rebuild. If you compile for 4.4, it will work with all later versions.

This is definitely not true.  Android changes requirements between API levels; native code compiled for API 19 (text relocations not enforced) will fail on  API 23 (text relocations enforced).  There are other examples of this behavior...see https://android-developers.googleblog.com/2016/06/android-changes-for-ndk-developers.html

If possible it would be good if the script could prompt for the target API level.

**Comment by qmfrederik** [comment]
@cydhaselton I'll update the script so that you can pass the API level as a parameter; I'll take 19 as the default and we can test whether the resulting coreclr is portable across API levels or not.

**Comment by cydhaselton** [comment]
@qmfrederik:  FYI, unpacking the Termux packages still fail; running the 
```
CONFIG_DIR=`realpath cross/android/arm64` ROOTFS_DIR=`realpath cross/android/toolchain/arm64/sysroot` ./build.sh cross arm64 skipgenerateversion skipmscorlib cmakeargs -DENABLE_LLDBPLUGIN=0
```
 fails due to missing `libicu`

**Comment by qmfrederik** [comment]
@cydhaselton dotnet/coreclr#9175 contains an updated script, it should be available in coreclr master soon.

**Comment by qmfrederik** [comment]
@cydhaselton I just realized that liblldb is available for Android as part of the lldb server ADK package. If you use my rootfs scripts, you'll find `liblldb.so.40` in `cross/android-rootfs/lldb/lib` 

The headers are missing but I'm thinking we can fetch them from here: https://android.googlesource.com/platform/external/lldb/+/android-6.0.0_r41/include/

**Comment by cydhaselton** [comment]
@qmfrederik I should have time to work on this today and tomorrow; let me know which branch and script I should use for the build and/or where you;ve posted build instructions.

**Comment by cydhaselton** [comment]
@qmfrederik Additionally, since you've made your own fork and PR based off of the work done here, you should open a new issue there for continuing work and reference this thread.

**Comment by qmfrederik** [comment]
@cydhaselton FYI - I've compiled the krb5 packages for Termux, you can find them here: 
- https://github.com/qmfrederik/dotnet-alpine-arm/raw/master/krb5-dev_1.15_aarch64.deb
- https://github.com/qmfrederik/dotnet-alpine-arm/raw/master/krb5_1.15_aarch64.deb

I didn't have the chance to test them yet; you should be able to download & install them in RootFS the same way as the other packages; this may help you cross compile CoreFX

As to the branches, think the best way for now is to check out the coreclr master and manually import the pull requests - life will be easier in a couple of days when all the PRs made it into master.

As to the issue & referencing, do you suggest we continue this in the coreclr repo or do you want me to create an issue in my repo? I think I tagged you in all the PRs so let me know if I missed something.

**Comment by cydhaselton** [comment]
@janvorli Sorry you had to do the krb work; I would've sent you the package I compiled. If you run into another prereq when building corefx post here; I've probably got the package.

Regarding issue/refs, since you;ve cherry-picked the commits to get Android building coreclr and since I'm still on a fairly steep learning curve with git (for example I'll need to figure out how to import your pull requests) I;m not sure where is the best place to create the issue. For now, if you have tag permissions in coreclr, create the issue there, reference this one, tag it with `port`, `android` and tag me.

 I'll  prep the local build tree (either by creating a new fork of master or an updated feature branch in my existing fork if possible) run the build, and any issues I run into I'll update your issue. If it's an issue for which I have a fix, I'll submit commits &| pull requests.

tagging @janvorli to review the above proposal.

**Comment by cydhaselton** [comment]
@janvorli Additionally, given the above info and @qmfrederik's work on the Android cross-build, would it be better to do the following:

1. Update my fork and feature-android branch from upstream master
2. Pull pending android-related PRs  into feature-android
3. Continue work in feature-android and update fork when PRs are merged.

**Comment by qmfrederik** [comment]
@cydhaselton My 2 cents: I think that proposal make sense, alternatively you can also wait for the PRs to be merged into master :)

Apart from that, I think the next two major work items are:
- Getting CoreFX to compile for .NET Core using the RootFS which we already built for CoreCLR. The first step would be to identify the missing packages. 
  * If you have a list of packages which we need whcih aren't in Termux yet, that would already be a great help.
   * If you have pre-compiled versions of those packages, that would be even better :)
   * I spent some time exploring how Termux does packaging and I have to admit, I'm a fan. So I'd be happy to assist in submitting PRs for those packages to be integrated in Termux. 
   * As a sidenote, now that we're cross-building CoreCLR from Linux, I'm also convinced it'll be a lot easier to get CoreCLR packages as a Termux package.
- Getting the LLDB plug in for .NET Core to build for arm64 (ideally for Android, Alpine & Ubuntu), see  dotnet/coreclr#8927 for that work

**Comment by janvorli** [comment]
@cydhaselton actually, looking at the changes in your fork / branch and given the changes that I have already cherry picked, I wonder which changes besides the temp folder path change are really needed. Unless I am missing something, changes in the msbuild related files look like hacks that you were using at various stages and that should not be needed. There seem to be few little changes that probably will still be needed.
So it might be easier as a new starting point to update the master in your fork from the upstream master, create a new branch and cherry pick or hand port only the necessary changes instead of spending time with rebasing your original change on the master and resolving many conflicts.

**Comment by cydhaselton** [comment]
@qmfrederik:  Depending on how long it takes for the script stuff to resolve, I may wait on PRs to be merged as I'm currently swamped at work.

Regarding building Android for CoreFX, check the feature-android branch [here](https://github.com/cydhaselton/cfx-android/tree/feature-android).  It needs to be updated from master...which I'll do when I have a minute...but I've made commits that will allow building on Android.

> I spent some time exploring how Termux does packaging and I have to admit, I'm a fan. So I'd be happy to assist in submitting PRs for those packages to be integrated in Termux.

There's an open issue/feature-request for dotnet in Termux; it would be fantastic if you could help with that effort as it'll take some cycles for me to get up to speed with their build scripts.

@janvorli:  

> So it might be easier as a new starting point to update the master in your fork from the upstream master, create a new branch and cherry pick or hand port only the necessary changes instead of spending time with rebasing your original change on the master and resolving many conflicts.

Sounds like a plan, once PRs are merged.

**Comment by cydhaselton** [comment]
@qmfrederik:  cfx-android/master [updated from](https://github.com/cydhaselton/cfx-android/pull/1) upstream master and cfx-android/feature-android  [updated from](https://github.com/cydhaselton/cfx-android/pull/2) cfx-android/master.

Reminder:  still on a steep learning curve with git; there could be problem with commits, PRs and/or merges within the cfx-android fork.  If reporting a problem, tag me and provide tips or links with info on fixing it and I'll do clean up.

**Comment by qmfrederik** [comment]
@cydhaselton To keep your for up to date with coreclr, there's an easier way without using PRs; this will also avoid the merge commits GitHub may generate for you.

```
git clone https://github.com/cydhaselton/coreclr/
cd coreclr
git remote add upstream https://github.com/dotnet/coreclr/
git fetch upstream
```

Because your fork and upstream have diverged, you'll need to reset your master to point to upstream/master:
```
git checkout master
git reset --hard upstream/master
git push --force
```

After that, any time you want to update:
```
git checkout master
git fetch upstream
git merge upstream/master
git push
```

To create a branch with all my PRs applied:

```
git remote add qmfrederik https://github.com/qmfrederik/coreclr
git fetch qmfrederik
git checkout master
git checkout -b android-all-prs
git pull [name of my branch, as shown in PR1]
git pull [name of my branch, as shown in PR2]
git pull [name of my branch, as shown in ...]
git push
```

**Comment by cydhaselton** [comment]
@qmfrederik Thanks.  I'll attend to that today.

Regarding CoreFX build and dependencies..

1. For dependencies not provided by Termux that I built on-device I'll do my best to package them with their build scripts.  Since that may take a while, I'll try to get the tarballs posted somewhere where you can download them today.

2. For dependencies provided by Termux (i.e. ones I installed to build CoreFX on device) you'll need the [libzopfli-dev](http://termux.net/dists/stable/main/binary-aarch64/libzopfli-dev_1.0.1-1_aarch64.deb) and possibly [libzopfli ](http://termux.net/dists/stable/main/binary-aarch64/libzopfli_1.0.1-1_aarch64.deb)packages; I used them in place of zlib (see [here](https://github.com/dotnet/corefx/compare/master...cydhaselton:feature-android#diff-687b2289579b8c5ddbb7aaa5dc6fde0d))

3. I'll clean up the commits I made to get CoreFX to build on Android (for example [this](https://github.com/dotnet/corefx/compare/master...cydhaselton:feature-android#diff-314c961c111ec07d658ef1db261bc042) and [this](https://github.com/dotnet/corefx/compare/master...cydhaselton:feature-android#diff-fad3610e964c713e5a35c57d076b2e38)) as soon as possible.

**Comment by cydhaselton** [comment]
@qmfrederik Finally updated fork and am ready to pull in PRs to a new branch. Quick question: is there a way to show/list all PRs for a branch?

**Comment by cydhaselton** [comment]
@qmfrederik Had some time so I updated my fork but I'm having some difficulties pulling in the PRs from your branch. 

Your instructions state:
```
git pull [name of my branch, as shown in PR1]
```

I've tried `git pull` with a number of variations of your branch names, but all fail with the following error:

```
ubuntu@ip-172-31-15-196:/build/core/coreclr$ git pull https://github.com/qmfrederik/coreclr/tree/android/build-scripts
remote: Not Found                                                                           
fatal: repository 'https://github.com/qmfrederik/coreclr/tree/android/build-scripts/' not found                                                                                         ubuntu@ip-172-31-15-196:/build/core/coreclr$ git pull https://github.com/qmfrederik/coreclr/android/build-scripts                                                                       remote: Not Found                                                                           
fatal: repository 'https://github.com/qmfrederik/coreclr/android/build-scripts/' not found
```

**Comment by qmfrederik** [comment]
@cydhaselton Assuming you already did: 

```
git remote add qmfrederik https://github.com/qmfrederik/coreclr
git fetch qmfrederik
```

I think you should be able to do 

```
git pull android/build-scripts/
```

or

```
git pull qmfrederik/android/build-scripts/
```

So the name of the branch is `android/build-scripts/`, not the full URL.

**Comment by cydhaselton** [comment]
@qmfrederik I did add your fork as a remote:

```
ubuntu@ip-172-31-15-196:/build/core/coreclr$ git remote -vv                                 
origin  https://github.com/cydhaselton/coreclr.git (fetch)                                  
origin  https://github.com/cydhaselton/coreclr.git (push)                                   
qmfrederik      https://github.com/qmfrederik/coreclr (fetch)                               
qmfrederik      https://github.com/qmfrederik/coreclr (push)                                
upstream        https://github.com/dotnet/coreclr/ (fetch)                                  
upstream        https://github.com/dotnet/coreclr/ (push)
```

Unfortunately neither command works:
```
ubuntu@ip-172-31-15-196:/build/core/coreclr$ git pull android/build-scripts/                
fatal: 'android/build-scripts/' does not appear to be a git repository                      
fatal: Could not read from remote repository. 
Please make sure you have the correct access rights                                         
and the repository exists.
```

I get the same error when using `qmfrederik/android/build-scripts`

**Comment by cydhaselton** [comment]
@qmfrederik This worked:

```
ubuntu@ip-172-31-15-196:/build/core/coreclr$ git pull qmfrederik refs/heads/android/toolchain-scripts                                                                                  
From https://github.com/qmfrederik/coreclr                                                   
* branch            android/toolchain-scripts -> FETCH_HEAD                                
Merge made by the 'recursive' strategy.                                                      
.gitignore                        |   1 +                                                   
Documentation/building/android.md | 102 +++++++++++++++++++++++++++++++++++                 
cross/build-android-rootfs.sh     | 145 ++++++++++++++++++++++++++++++++++++++++++++++++++  
3 files changed, 248 insertions(+)
```

Which of the following should I pull for cross-building android?

```
ubuntu@ip-172-31-15-196:/build/core/coreclr$ git ls-remote qmfrederik
ab709855f21ca49a63b148d5c4e0be5e3b8b5e28   HEAD
bf11611174209665d55532eb9c97861e89b5de11        refs/heads/SpanOfT
820c939d05cac2afcb4d11092ea9ae355c72a91e        refs/heads/android/build-scripts
597a7c02069d9ccd6e358c8eeaf4eaae75e875e9        refs/heads/android/common-test-libraries
e8ccd261a9bfdf4fc0ac3fed294070f4f73d1567        refs/heads/android/cross
8f213286486bfe59c3dd6d188a75a4ce081e0cb4        refs/heads/android/duplicate-definitions
c4630a89ad4458e844f699e02d030bc43bcb45eb        refs/heads/android/pal_function_cleanup
38a71e9fde512bdb88bb29d4464d5224aed99ef4        refs/heads/android/temp-path
4aea198d1d86b39529eaee16f711c98c88147808        refs/heads/android/toolchain-scripts
4bf6d083099f1fb99750763ffa442f4842b7bbef        refs/heads/features/alpine-arm-64
bcbd08f825b6a3f873fc279380c1c00d3545fe19        refs/heads/features/android
31d07b5fd5f288d87fef04e809542c5b87cc0271        refs/heads/features/android-rebased
2573063b3dec3d8069102acf5c91995b12fdd882        refs/heads/features/android-threading
```

**Comment by qmfrederik** [comment]
You'll need at least these:

* android/duplicate-definitions
* android/cross
* android/toolchain-scripts

These may also be of interest (f you do a git fetch of my repo to refresh the refs you have)
* port-121d095
* arm64/port-5821
* arm64/port-5339

**Comment by cydhaselton** [comment]
@qmfrederik Hmm. I'm getting the following error during configure (post-script)

```
-- Performing Test HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
-- Performing Test HAVE_FULLY_FEATURED_PTHREAD_MUTEXES - Failed
CMake Error: Error in cmake code at                                                         /build/core/coreclr/src/pal/tests/palsuite/CMakeLists.txt:8:                                
Parse error.  Expected a command name, got unquoted argument with text "<<<<<<<".           -- Configuring incomplete, errors occurred! 
See also "/build/core/coreclr/bin/obj/Linux.arm64.Debug/CMakeFiles/CMakeOutput.log".        
See also "/build/core/coreclr/bin/obj/Linux.arm64.Debug/CMakeFiles/CMakeError.log".         /build/core/coreclr
Failed to generate CoreCLR component build project!
```

**Comment by cydhaselton** [comment]
@qmfrederik In which PR did you put the `lpthread` and `lrt` changes?

```
/build/core/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find -lpthread
clang38++: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: ***[src/pal/tests/palsuite/exception_handling/pal_sxs/test1/libpaltest_pal_sxs_test1_dll2.so] Error 1                                                                         make[1]: *** [src/pal/tests/palsuite/exception_handling/pal_sxs/test1/CMakeFiles/paltest_pal_sxs_test1_dll2.dir/all] Error 2                                                            
make: *** [all] Error 2
```

**Comment by qmfrederik** [comment]
@cydhaselton That should have beem https://github.com/dotnet/coreclr/pull/9173, which has been merged into master already. The arm64-specific fixes also made it into master, which makew me think we're closer to getting Hello World to run on Android.

**Comment by cydhaselton** [comment]
@qmfrederik About damn time

Regarding the PR, I did pull that but there was a conflict for the `src/pal/tests/palsuite/CMakeLists.txt` file. I'll re-try the pull, and if that doesn;t work I'll change it manually

**Comment by cydhaselton** [comment]
@qmfrederik CMakeFiles are updated but build still failing…was there a PR to set -DANDROID or -D__ANDROID__?

**Comment by qmfrederik** [comment]
@cydhaselton I don't think so; I think `__ANDROID__` should be set automatically by the compiler although I've noticed that does not always appear to be the case.

**Comment by cydhaselton** [comment]
@qmfrederik The CMakeLists.txt files have a conditional for skipping `lpthread` and `lrt`
```
if(NOT CLR_CMAKE_PLATFORM_ANDROID)
…
```
What needs to be set in order for this to detect ANDROID?

**Comment by cydhaselton** [comment]
@qmfrederik Worked around the lpthread issue by creating a soft link, but now I'm running into the `reserved` issue. Was there a PR for that?  I applied `duplicate-definitions` already.

**Comment by qmfrederik** [comment]
@cydhaselton The CLR_CMAKE_PLATFORM_ANDROID flag is set in `toolchain.cmake` - see https://github.com/dotnet/coreclr/pull/9174/files#diff-258b45c0463fafa563c517ce3cba61daR3

Regarding the reserved issue, I don't think I hit it the last time I compiled CoreCLR (i.e. a separate PR wasn't needed), but I need to check.

**Comment by qmfrederik** [comment]
@cydhaselton PRs are also flowing into CoreFX to get the cross build for Android up and running
https://github.com/dotnet/corefx/pulls?utf8=%E2%9C%93&q=is%3Apr%20is%3Aclosed%20%20label%3Aos-unsupported%20

The disadvantage of PRs is that all the work gets split up in small chunks and it takes some time to digest them. Once the CoreFX PRs are through, I'll start again from CoreCLR master + CoreFX master and do a double check to see whether everything compiles OK.

**Comment by vishalbiswas** [comment]
This may be useful. I have a pull request https://github.com/termux/termux-packages/pull/742 for llvm which includes lldb (lldb builds successfully). But I haven't been able to get it working. It fails when trying to execute `start` in lldb prompt.

**Comment by cydhaselton** [comment]
@qmfrederik Made some changes to resolve the `__reserved` issue and the build completes. Restoring tools fails due to lack of space on my EC2 instance; I'll need to take some time to expand the drive or start over.

**Comment by qmfrederik** [comment]
@cydhaselton That's great news!

I've also got CoreFX to fully cross-compile for Android. You can either import all open PRs on CoreFX, or start from the android/master branch in my CoreFX repo.

Let me have a look again at CoreCLR.

**Comment by qmfrederik** [comment]
@cydhaselton You're right, I think the `__reserved` changes are still needed and they didn't make it into a PR.

**Comment by cydhaselton** [comment]
@qmfrederik Pushed `__reserved` changes to fork; they may need some revision which I'll tackle after re-building my EC2 instance

**Comment by cydhaselton** [comment]
EC2 instance rebuilt; working on getting local repo set up

@qmfrederik have you had a chance to look at the `__reserved` commits?

**Comment by qmfrederik** [comment]
@cydhaselton The `__reserved` issue should now be fixed in coreclr/master, since https://github.com/dotnet/coreclr/pull/9402

**Comment by cydhaselton** [comment]
@qmfrederik: Good to know; will update fork and re-run.

Going forward it'd be helpful if you let me submit the PR requests and make changes if necessary so that I can gain proficiency with the coreclr workflow. If time is a factor let me know and I'll adjust my priorities accordingly.

**Comment by cydhaselton** [comment]
@qmfrederik Set up local repo, updated, ran ./cross/build-android-rootfs.sh then the build command and ran into the following errors: http://pastebin.com/eEwrWxhm
